x-pack/plugins/endpoint/server/routes/resolver/children.ts

```diff
@@ -41,46 +42,90 @@ export const validateChildren = {
     after: schema.maybe(schema.string()),
     limit: schema.number({ defaultValue: 10, min: 1, max: 100 }),
     legacyEndpointID: schema.maybe(schema.string()),
+    levels: schema.number({ defaultValue: 0, min: 0, max: 3 }),
   }),
 };
 
+async function potentiallyKillKibana(
+  client: IScopedClusterClient,
+  id: string,
+  legacyEndpointID: string | undefined,
+  pagination: PaginationParams,
+  levels: number
+): Promise<any> {
+  const childrenQuery = new ChildrenQuery(legacyEndpointID, pagination);
+  const lifecycleQuery = new LifecycleQuery(legacyEndpointID);
+
+  // Retrieve the related child process events for a given process
+  const { total, results: events, nextCursor } = await childrenQuery.search(client, id);
+  const childIDs = events.map(extractEntityID);
+
+  // Retrieve the lifecycle events for the child processes (e.g. started, terminated etc)
+  // this needs to fire after the above since we don't yet have the entity ids until we
+  // run the first query
+  const { results: lifecycleEvents } = await lifecycleQuery.search(client, ...childIDs);
+
+  // group all of the lifecycle events by the child process id
+  const lifecycleGroups = _.groupBy(lifecycleEvents, extractEntityID);
+  const children = await Promise.all(
+    Object.entries(lifecycleGroups).map(async ([entityID, group]) => {
+      // Because this is recursive and in a map call, this is going to issue a maximum of
+      // (2 * (m - 1)) ^ n queries where n is the level of children to display and m is the limit of
+      // child processes to query per level. Each of these is going to be held in memory and reconstructed
+      // so if we have large documents this is going to be slow and hog a ton of memory. Rather than designing
```

Comment:
❔ Would it reduce the pressure on memory somewhat if instead of glomming it all together in one `res.ok` we flush it through a WritableStream at each step. Something like 
`
res.write(childrenFromThisLevel)
breakKibana(levels -1)
`

Reply:
Using some kind of WritableStream mechanism, could you also "stop" or "pause" flushing into it if it hits some kind of highWaterMark for backpressure so if it isn't getting emptied... I know the core Node stuff has some facilities for that, and I think they are fond of RxJS here, which I don't know much about but I imagine it has some things for that, too, maybe.